spec/requests/profile_spec.rb

```diff
@@ -0,0 +1,166 @@
+        let(:profile_owner) { build_stubbed(:user2) }
+
+        before do
+          allow(User).to receive(:find_by).and_return(profile_owner)
+          get profile_index_path, params: {uid: profile_owner.uid}
+        end
+
+        it "does not have the Stories section" do
+          expect(response.body).to_not include("Stories")
+        end
+
+        it "returns the profile for the owner" do
+          expect(response.body).to include(profile_owner.name)
+        end
+      end
+    end
+  end
+
+  describe "#data" do
+    let(:user) { create(:user) }
+    let!(:moment) { create(:moment, user: user) }
+
+    context "when the user is signed in" do
+      before { sign_in user }
+      before { get data_profile_index_path, params: {page: 1, uid: user.uid}, headers: {"ACCEPT" => "application/json"} }
+
+      it "returns a response with the correct path" do
+        expect(JSON.parse(response.body)["data"].first["link"]).to eq moment_path(moment)
+      end
+    end
+
+    context "when the user is not signed in" do
+      before { get data_profile_index_path params: {page: 1, uid: user.uid}, headers: {"ACCEPT" => "application/json"} }
+
+      it "returns a no_content status" do
+        expect(response).to have_http_status(:no_content)
+      end
+    end
+  end
+
+  describe "#add_ban" do
+    let(:user) { create(:user2) }
+
+    before(:each) { Devise.mailer.deliveries.clear }
+
+    context "when admin does not exist" do
+      let(:non_admin_user) { create(:user1) }
+
+      it "cannot ban user" do
+        sign_in non_admin_user
+        post add_ban_profile_index_path, params: {user_id: user.id}
+        expect(response).to have_http_status(204)
+        expect(Devise.mailer.deliveries.count).to eq(0)
+      end
+    end
+
+    context "when admin exists" do
+      let(:admin_user) { create(:user1, admin: true) }
+
+      context "when user exists" do
+        it "bans the user" do
+          sign_in admin_user
+          expect { post add_ban_profile_index_path, params: {user_id: user.id} }
+            .to change(Devise.mailer.deliveries, :count).by 1
+          expect(response).to redirect_to(admin_dashboard_path)
+          expect(flash[:notice]).to eq("#{user.name} has been banned")
+        end
+      end
+
+      context "when user does not exist" do
+        it "does not ban the user" do
+          sign_in admin_user
+          expect { post add_ban_profile_index_path, params: {user_id: -1} }
+            .not_to change(Devise.mailer.deliveries, :count).from(0)
+          expect(response).to redirect_to(admin_dashboard_path)
+          expect(flash[:alert]).to eq("Could not ban -1")
+        end
+      end
+    end
+  end
+
+  describe "#remove_ban" do
+    let(:banned_user) { create(:user2, banned: true) }
+
+    before(:each) { Devise.mailer.deliveries.clear }
+
+    context "when admin does not exist" do
+      let(:nonadmin_user) { create(:user1) }
+
+      it "cannot ban user" do
+        sign_in nonadmin_user
```

Comment:
for DRYness purposes, it would be better to move the `sign_in` to a `before { }` block. (applies to all `sign_in` calls in `it` blocks)

Reply:
Definetly agree for the 'when signed in user is an admin' context but given there is only one test in the 'when signed in user is not an admin' do you think that's needed? Certainly if we were to add more tests

Reply:
I usually go by a rule of 3 for myself. This is fine for now to me although with `sign_in` I can see it be intuitive to have it there in the first place in the context block. On your digression I'd say 🤷

Reply:
On the other hand, I noticed there are some `sign_in` inside context and others inside `it` so for consistency, I'd move it in the context block.